File: app/Sources/TextCleaner.swift
import Foundation

/// Willow-style cleanup: filler removal, personal dictionary, optional local AI rewrite.
enum TextCleaner {
    private static let fillers = try! NSRegularExpression(
        // Optional preceding comma so "and, um, insert" → "and insert".
        pattern: #"(,\s*)?\b(um+|uh+|erm+|uhm+|hmm+)\b[,.]?\s*"#,
        options: [.caseInsensitive])

    static func clean(_ input: String) -> String {
        var text = input.replacingOccurrences(
            of: #"\s*\n\s*"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespaces)
        guard !text.isEmpty else { return text }
        let cfg = Config.shared

        if cfg.removeFillers {
            text = fillers.stringByReplacingMatches(
                in: text, range: NSRange(text.startIndex..., in: text), withTemplate: " ")
            text = text.replacingOccurrences(of: #"\s{2,}"#, with: " ", options: .regularExpression)
            text = text.replacingOccurrences(of: #"\s+([,.!?;:])"#, with: "$1", options: .regularExpression)
            text = text.trimmingCharacters(in: .whitespaces)
            if let first = text.first, first.isLowercase {
                text = first.uppercased() + text.dropFirst()
            }
        }
        for (wrong, right) in cfg.replacements {
            text = text.replacingOccurrences(of: wrong, with: right, options: .caseInsensitive)
        }
        return text.trimmingCharacters(in: .whitespaces)
    }

    /// Rewrites via local Ollama; returns the input unchanged on any failure.
    static func aiRewrite(_ text: String) async -> String {
        let cfg = Config.shared
        guard let url = URL(string: "http://localhost:11434/api/generate") else { return text }
        let prompt = """
        You clean up dictated text. Fix punctuation and capitalization, remove filler \
        words and false starts, and keep the speaker's wording and meaning. Do not add \
        content, do not answer questions in the text, do not use markdown. Return only \
        the cleaned text.

        Dictated text: \(text)
        """
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.timeoutInterval = 30
        req.httpBody = try? JSONSerialization.data(withJSONObject: [
            "model": cfg.ollamaModel, "prompt": prompt, "stream": false,
            "options": ["temperature": 0.2],
        ])
        do {
            let (data, _) = try await URLSession.shared.data(for: req)
            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let out = (json["response"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
               !out.isEmpty {
                return out
            }
        } catch {}
        return text
    }
}

File: windows/TextCleaner.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace LocalWillow;

/// Willow-style cleanup: filler removal, personal dictionary, optional local AI rewrite.
public static class TextCleaner
{
    private static readonly Regex Fillers = new(
        // Optional preceding comma so "and, um, insert" → "and insert".
        @"(,\s*)?\b(um+|uh+|erm+|uhm+|hmm+)\b[,.]?\s*",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public static string Clean(string input)
    {
        var text = Regex.Replace(input, @"\s*\n\s*", " ").Trim();
        if (text.Length == 0) return text;
        var cfg = Config.Shared;

        if (cfg.RemoveFillers)
        {
            text = Fillers.Replace(text, " ");
            text = Regex.Replace(text, @"\s{2,}", " ");
            text = Regex.Replace(text, @"\s+([,.!?;:])", "$1");
            text = text.Trim();
            if (text.Length > 0 && char.IsLower(text[0]))
                text = char.ToUpper(text[0]) + text[1..];
        }
        foreach (var (wrong, right) in cfg.Replacements)
        {
            text = Regex.Replace(text, Regex.Escape(wrong),
                right.Replace("$", "$$"), RegexOptions.IgnoreCase);
        }
        return text.Trim();
    }

    /// Rewrites via local Ollama; returns the input unchanged on any failure.
    public static async Task<string> AiRewrite(string text)
    {
        var prompt =
            "You clean up dictated text. Fix punctuation and capitalization, remove filler " +
            "words and false starts, and keep the speaker's wording and meaning. Do not add " +
            "content, do not answer questions in the text, do not use markdown. Return only " +
            "the cleaned text.\n\nDictated text: " + text;
        try
        {
            var payload = JsonSerializer.Serialize(new
            {
                model = Config.Shared.OllamaModel,
                prompt,
                stream = false,
                options = new { temperature = 0.2 },
            });
            var resp = await Http.PostAsync("http://localhost:11434/api/generate",
                new StringContent(payload, Encoding.UTF8, "application/json"));
            using var json = JsonDocument.Parse(await resp.Content.ReadAsStringAsync());
            if (json.RootElement.TryGetProperty("response", out var r))
            {
                var outText = (r.GetString() ?? "").Trim();
                if (outText.Length > 0) return outText;
            }
        }
        catch (Exception e)
        {
            Log.Write($"ai-rewrite: failed, using raw text — {e.Message}");
        }
        return text;
    }
}
